synedrion/src/www02/bip32.rs

```diff
@@ -0,0 +1,206 @@
+    VerifyingKey<P::Curve>: PubTweakable,
+{
+    fn derive_verifying_key_bip32(
+        &self,
+        derivation_path: &DerivationPath,
+    ) -> Result<VerifyingKey<P::Curve>, bip32::Error> {
+        let public_key = self.verifying_key().map_err(|_| bip32::Error::Crypto)?;
+        let tweakable_pk = public_key.tweakable_pk();
+        let tweaks = derive_tweaks::<<P as SchemeParams>::Curve>(&tweakable_pk, derivation_path)?;
+        apply_tweaks_public(tweakable_pk, &tweaks)
+    }
+}
+
+impl<C> DeriveChildKey<C> for VerifyingKey<C>
+where
+    C: CurveArithmetic + PrimeCurve + DigestPrimitive,
+    <C as CurveArithmetic>::AffinePoint: FromEncodedPoint<C> + ToEncodedPoint<C>,
+    <C as Curve>::FieldBytesSize: ModulusSize,
+    VerifyingKey<C>: PubTweakable,
+{
+    fn derive_verifying_key_bip32(&self, derivation_path: &DerivationPath) -> Result<VerifyingKey<C>, bip32::Error> {
+        let tweakable = self.tweakable_pk();
+        let tweaks = derive_tweaks::<C>(&tweakable, derivation_path)?;
+        apply_tweaks_public(tweakable, &tweaks)
+    }
+}
+
+fn derive_tweaks<C>(
+    public_key: &(impl bip32::PublicKey + Clone),
+    derivation_path: &DerivationPath,
+) -> Result<Vec<PrivateKeyBytes>, bip32::Error>
+where
+    C: CurveArithmetic + PrimeCurve + DigestPrimitive,
+    <C as CurveArithmetic>::AffinePoint: FromEncodedPoint<C> + ToEncodedPoint<C>,
+    <C as Curve>::FieldBytesSize: ModulusSize,
+{
+    let mut public_key = public_key.clone();
+
+    // Note: deriving the initial chain code from public information. Is this okay? <–– PROBABLY NOT OK?
+    let pubkey_bytes = public_key.to_bytes();
+    let initial_chain_code = <C as DigestPrimitive>::Digest::new()
+        .chain_update(b"chain-code-derivation")
+        .chain_update(pubkey_bytes)
+        .finalize();
+    let mut chain_code: ChainCode = Default::default();
+    let len = initial_chain_code.len().min(chain_code.len());
+    chain_code
+        .get_mut(..len)
+        .ok_or(bip32::Error::Decode)?
+        .copy_from_slice(initial_chain_code.get(..len).ok_or(bip32::Error::Decode)?.as_ref());
+
+    let mut tweaks = Vec::new();
+    for child_number in derivation_path.iter() {
+        let (tweak, new_chain_code) = public_key.derive_tweak(&chain_code, child_number)?;
+        public_key = public_key.derive_child(tweak)?;
+        tweaks.push(tweak);
+        chain_code = new_chain_code;
+    }
+
+    Ok(tweaks)
+}
+
+fn apply_tweaks_public<C>(
+    public_key: impl bip32::PublicKey + Clone,
+    tweaks: &[PrivateKeyBytes],
+) -> Result<VerifyingKey<C>, bip32::Error>
+where
+    C: CurveArithmetic + PrimeCurve,
+    <C as Curve>::FieldBytesSize: ModulusSize,
+    <C as CurveArithmetic>::AffinePoint: FromEncodedPoint<C> + ToEncodedPoint<C>,
+{
+    let mut public_key = public_key;
+    for tweak in tweaks {
+        public_key = public_key.derive_child(*tweak)?;
+    }
+    let offset = bip32::KEY_SIZE - <C as Curve>::FieldBytesSize::USIZE;
```

Comment:
I'm not sure I understand how this works. `bip32::KEY_SIZE = 32`, `<C as Curve>::FieldBytesSize::USIZE` is 32 for `k256` or 24 for `tiny-curve`. Then the length of `public_key.to_bytes()` is 33, with all of these bytes being meaningful for `k256`. Then we just drop the first byte (the sign of `y`) and it works as a SEC1 representation?

Reply:
Also the question above about unwrapping applies here as well - I don't think you need to de/serialize at all.

Reply:
Let me spell out my reasoning and you can shoot it down!

- For `k256` the offset is `0`, so `bytes.get(offset..)` will be the whole slice of 33 bytes yes? Should behave the same as the old code.
- For `TinyCurve64` the offset is `8`, so in this case we truncate the high 8 bytes. This is likely not ok for a proper SEC1 repr, but the premise here is that BIP32 isn't really defined for anything other than secp256k1 so it's unclear what "valid" would even mean.

Reply:
Hm, perhaps it does work, but in any case it's not necessary - we can avoid serialization entirely, see the other comment.